### ISSUE TITLE
TypeError: Cannot read property 'resolve' of undefined

### ISSUE THREAD

**Issue by dessaya** [NatalieWolfe/node-promise-pool#5]
Hi! First of all, thanks for the package :)

We are using it to provide a pool of redis connections, using something like this code:

```js
import bluebird from 'bluebird';
import redis from 'redis';
bluebird.promisifyAll(redis.RedisClient.prototype);
bluebird.promisifyAll(redis.Multi.prototype);

import pool from 'generic-promise-pool';

export default {
  createPool: () => {
    return pool.create({
      name: 'redis',
      max: 10,
      min: 1,

      create: () => new Promise((resolve, reject) => {
        const conn = redis.createClient(config.server);
        conn.once('ready', () => resolve(conn));
        conn.once('error', e => reject(e));
      }),

      destroy: conn => new Promise((resolve, reject) => {
        conn.quit();
        conn.once('end', resolve);
      }),
    });
  },
};
```

This seems to mostly work in production, but we are occasionally getting this error:

```
TypeError: Cannot read property 'resolve' of undefined
at /home/app/brainiac/node_modules/generic-promise-pool/lib/PromisePool.js:341:47
at wrappedPromise.linkTransaction (/home/app/brainiac/node_modules/newrelic/lib/instrumentation/core/globals.js:183:23)
at wrappedPromise.wrapped [as __NR_wrapper] (/home/app/brainiac/node_modules/newrelic/lib/transaction/tracer/index.js:182:28)
at wrappedHandler (/home/app/brainiac/node_modules/newrelic/lib/instrumentation/core/globals.js:217:26)
at process._tickDomainCallback [as _tickCallback] (internal/process/next_tick.js:129:7)
```

Any idea about what might be causing this? Perhaps we are not properly releasing the connections? Or we are not handling some error? Or is it some bug in the package?

After looking at PromisePool.js and PriorityQueue.js, I think the only way this error might be raised is after calling `PriorityQueue.dequeue()` when the queue is empty. But that's all I got, I don't see how this can happen in PropisePool.

Thanks!

**Comment by NatalieWolfe** [comment]
Thanks for reporting this! I'm also glad to see the New Relic promise instrumentation in your stack trace, I wrote that too. :)

I just got back from a vacation, so I haven't had a chance to reproduce this yet, but I have a theory it might actually have to do with your `create` function. I think you might be inadvertently calling `reject` on a resolved promise if the connection ever errors after being created. Can you try changing it to this?

```js
      create: () => new Promise((resolve, reject) => {
        const conn = redis.createClient(config.server);
        conn.once('ready', () => {
          conn.removeListener('error', reject)
          resolve(conn)
        });
        conn.once('error', reject);
      })
```

I'll also try to repro this tonight and will get back to you once I have a fix.

**Comment by dessaya** [comment]
Oops! I should have seen that... I'll try your suggestion and let you know if it fixes the error (I have to test it for a few hours as the error appears only occasionally).

PS: Thanks for your work in node-newrelic as well! :)

**Comment by dessaya** [comment]
Unfortunately the error keeps showing up in production :-(

**Comment by NatalieWolfe** [comment]
I've managed to reproduce the issue with the following steps:

1) Use a pool with min: 0, max: 2, and no resources allocated yet.
2) Acquire a connection (1).
3) Before releasing connection 1, acquire a second connection (2).
4) Before connection 2 is ready (i.e. before `.create` resolves its promise), release connection 1.
5) When connection 2 becomes ready there are no waiting clients and thus `.dequeue` returns `undefined`.

The fix for this is fairly easy, I'll just check that `.dequeue` actually returned something. Working on creating a test to prevent regressions right now.

**Comment by NatalieWolfe** [comment]
@dessaya If you can, I'd appreciate you trying the PR I just pushed up. If you set the "version" of the generic-promise-pool dependency in your package.js to `NatalieWolfe/node-promise-pool#resolve-undefined` as shown below you can deploy the branch directly. This fix should resolve the problem you're encountering, but I'd love to have positive confirmation before moving forward.

```json
{
  "name": "foobar",
  "version": "1.2.3",
  "dependencies": {
    "generic-promise-pool": "NatalieWolfe/node-promise-pool#resolve-undefined"
  },
}
```

**Comment by dessaya** [comment]
We pushed that version to production and the error has not come up so far. It seems to be fixed :)

Thanks!

**Comment by NatalieWolfe** [comment]
Excellent! I'll push out a new release this week with this fix. Thanks for helping make it better!

**Comment by NatalieWolfe** [comment]
Just released version 1.2.1 with this fix. Thanks again for reporting this to me!
